transform_constantfold.ml

```diff
@@ -1,111 +1,130 @@
+ * This optimization combines constant propagation, copy propagation, and
+ * constant folding.
  *
- * Then, whenever `x` is used in an expression (while x is still in scope), it
- * is replaced by the literal `l`. Afterwards, the variable `x` is no longer
- * used, and the declaration can be removed by running `minimize_lifetimes`.
+ * The strategy is to maintain a "propagation environment" mapping variables to
+ * simple expressions (constants or variables) and apply the `normalize`
+ * function to expressions.
+ *
+ * `normalize` will use the environment to perform as many substitutions as it
+ * can. If this results in a new constant definition, then that definition is
+ * added to the environment. `drop x` will remove `x` and its mapping from the
+ * environment.
  *)
-let const_prop ({formals; instrs} : analysis_input) : instructions option =
-  (* Finds the declarations that can be used for constant propagation.
-     Returns a list of (pc, x, l) where `const x = l` is defined at pc `pc`. *)
-  let rec find_candidates instrs pc acc =
-    if pc = Array.length instrs then acc
-    else match[@warning "-4"] instrs.(pc) with
-    | Decl_const (x, Simple(Constant(l))) ->
-        find_candidates instrs (pc+1) ((pc, x, l) :: acc)
-    | _ -> find_candidates instrs (pc+1) acc
+let const_fold (({formals; instrs} as inp) : analysis_input) : instructions option =
+  let module Prop_env = Map.Make(Variable) in
+
+  (* Perform constant and copy propagation on the expression `exp`, using
+   * the mappings in `penv`.
+   * Returns an updated expression. *)
+  let propagate exp penv =
+    let try_prop x e =
+      if Prop_env.mem x penv
+      then Edit.replace_var_in_exp x (Prop_env.find x penv) e else e in
+    VarSet.fold try_prop (expr_vars exp) exp
   in
 
-  (* Replaces the variable `x` with literal `l` in instruction `instr`. *)
-  let convert x l instr =
-    let replace = Edit.replace_var_in_exp x l in
-    let replace_arg = Edit.replace_var_in_arg x l in
+  (* Constant fold the expression `exp`.
+   * Returns an updated expression. *)
+  let fold exp =
+    match[@warning "-4"] exp with
+    | Simple _ -> exp
+    | Op (op, [Constant(v1); Constant(v2)]) ->
+      begin match op with
+      | Eq -> Simple (Constant (Bool (Eval.value_eq v1 v2)))
+      | Neq -> Simple (Constant (Bool (Eval.value_neq v1 v2)))
+      | Plus -> Simple (Constant (Int (Eval.value_plus v1 v2)))
+      end
+    | Op (_, _es) -> exp
+  in
+
+  (* Try to add a new mapping to the environment. `exp` is the simple expression
+   * being bound to the constant variable `x`. *)
+  let try_add x exp scope penv =
+    match exp with
+    | Simple (Constant v) -> Prop_env.add x (Constant v) penv
+    | Simple (Var v) ->
+      (* `exp` is a variable v, need to check if `v` is const or mut. *)
+      begin match scope with
+      | DeadScope -> penv
+      | Scope scope ->
+        try
+          (* If `v` is mut, we'll get an `Incomparable` exception. *)
```

Comment:
oh, now I understand. But couldn't you just wait for the next iteration? Because constant propagation will replace v with the actual constant first, then you can check afterwards. No need to look up the function!